src/context/defrag.rs

```diff
@@ -0,0 +1,227 @@
+    ///
+    /// Smaller keys, keys that do not implement `free_effort` or the global
+    /// defrag callback are not called in late-defrag mode. In those cases, a
+    /// call to this function will return [`Status::Err`].
+    ///
+    /// The cursor may be used by the module to represent some progress into the
+    /// module's data type. Modules may also store additional cursor-related
+    /// information locally and use the cursor as a flag that indicates when
+    /// traversal of a new key begins. This is possible because the API makes
+    /// a guarantee that concurrent defragmentation of multiple keys will
+    /// not be performed.
+    pub fn set_cursor(&self, cursor: u64) -> Status {
+        unsafe {
+            RedisModule_DefragCursorSet.expect("RedisModule_DefragCursorSet should be available.")(
+                self.defrag_ctx,
+                cursor,
+            )
+        }
+        .into()
+    }
+
+    /// Fetch a cursor value that has been previously stored using [`Self::set_cursor`].
+    /// If not called for a late defrag operation, [`Err`] will be returned.
+    pub fn get_cursor(&self) -> Result<u64, RedisError> {
+        let mut cursor: u64 = 0;
+        let res: Status = unsafe {
+            RedisModule_DefragCursorGet.expect("RedisModule_DefragCursorGet should be available.")(
+                self.defrag_ctx,
+                (&mut cursor) as *mut u64,
+            )
+        }
+        .into();
+        if res == Status::Ok {
+            Ok(cursor)
+        } else {
+            Err(RedisError::Str("Could not get cursor value"))
+        }
+    }
+
+    /// Defrag a memory allocation previously allocated by RM_Alloc, RM_Calloc, etc.
+    /// The defragmentation process involves allocating a new memory block and copying
+    /// the contents to it, like realloc().
+    ///
+    /// If defragmentation was not necessary, NULL is returned and the operation has
+    /// no other effect.
+    ///
+    /// If a non-NULL value is returned, the caller should use the new pointer instead
+    /// of the old one and update any reference to the old pointer, which must not
+    /// be used again.
+    ///
+    /// The function is unsafe because it is assumed that the pointer is valid and previusly
+    /// allocated. It is considered undefined if this is not the case.
+    pub unsafe fn defrag_realloc<T>(&self, mut ptr: *mut T) -> *mut T {
+        let new_ptr: *mut T = RedisModule_DefragAlloc
+            .expect("RedisModule_DefragAlloc should be available.")(
+            self.defrag_ctx, ptr.cast()
+        )
+        .cast();
+        if !new_ptr.is_null() {
+            ptr = new_ptr;
+        }
+        ptr
+    }
+
+    /// Allocate memory using defrag allocator if supported by the
+    /// current Redis server, fallback to regular allocation otherwise.
+    pub fn defrag_alloc<T>(&self, layout: Layout) -> *mut T {
+        unsafe { std::alloc::alloc(layout) }.cast()
```

Comment:
From docs:
```
This function forwards calls to the [GlobalAlloc::alloc](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) method of the allocator registered with the #[global_allocator] attribute if there is one, or the std crate’s default.
```
Global allocator is set on the scope of the entire ELF. This means it will call the global allocator as expected.

Reply:
Thanks.